test/tools/cli/IonProcess.java

```diff
@@ -0,0 +1,312 @@
+
+        IonProcess.ProcessArgs parsedArgs = new IonProcess.ProcessArgs();
+        CmdLineParser parser = new CmdLineParser(parsedArgs);
+        parser.getProperties().withUsageWidth(CONSOLE_WIDTH);
+        OutputFormat outputFormat = null;
+        String outputFileName = null;
+        String errorReportName = null;
+
+        try {
+            parser.parseArgument(args);
+            outputFormat = parsedArgs.getOutputFormat();
+            outputFileName = parsedArgs.getOutputFile();
+            errorReportName = parsedArgs.getErrorReport();
+        } catch (CmdLineException e) {
+            System.err.println(e.getMessage() + "\n");
+            System.err.println("\"Process\" reads the input file(s) and re-write in the format specified by --output.\n");
+            System.err.println("Usage:\n");
+            System.err.println("ion process [--output <file>] [--error-report <file>] [--output-format \n"
+                    + "(text | pretty | binary | events | none)] [--catalog <file>]... [--imports <file>]... \n"
+                    + "[--perf-report <file>] [--filter <filter> | --traverse <file>]  [-] [<input_file>]...\n");
+            System.err.println("Options:\n");
+            parser.printUsage(System.err);
+            System.exit(USAGE_ERROR_EXIT_CODE);
+        }
+
+        //Initialize output stream (default value: STDOUT)
+        OutputStream  outputStream = initOutputStream(outputFileName, SYSTEM_OUT_DEFAULT_VALUE);
+        //Initialize error report (default value: STDERR)
+        OutputStream errorReportOutputStream = initOutputStream(errorReportName, SYSTEM_ERR_DEFAULT_VALUE);
+
+        try (IonWriter ionWriterForOutput = outputFormat.createIonWriter(outputStream);
+             IonWriter ionWriterForErrorReport = outputFormat.createIonWriter(errorReportOutputStream);
+        ) {
+            if (parsedArgs.getOutputFormatName() == "events") {
+                processFilesInEvents(ionWriterForOutput, ionWriterForErrorReport, parsedArgs);
+            } else {
+                processFiles(ionWriterForOutput, ionWriterForErrorReport, parsedArgs);
+            }
+        } catch (IOException e) {
+            System.err.println("Failed to close OutputStream: " + e.getMessage());
+        }
+    }
+
+    private static void processFiles(IonWriter ionWriterForOutput,
+                                     IonWriter ionWriterForErrorReport,
+                                     IonProcess.ProcessArgs args) {
+        for (String path : args.getInputFiles()) {
+            try (
+                    InputStream inputStream = new BufferedInputStream(new FileInputStream(path));
+                    IonReader ionReader = IonReaderBuilder.standard().build(inputStream);
+            ) {
+                process(ionWriterForOutput, ionWriterForErrorReport, ionReader);
+
+                ionWriterForOutput.finish();
+                ionWriterForErrorReport.finish();
+            } catch (IOException e) {
```

Comment:
Since you catch IOException in main(), you no longer need to catch it here. Just declare that this method throws IOException.

Reply:
I handled most of the Exceptions in main now.

Question:
I moved OutputStream for output and errorReport inside try-with-resource now (IonProcess 62-65)

I noticed that try..catch method doesn't catch exceptions from the try block (IonProcess.java 70-74). For example, if processFilesInEvents method (IonProcess.java 73) throw an IOException, **catch(IOException e) { ... }** part (IonProcess.java line 75) isn't called.

I found that the reason of it is because of declaration of errorReportOutputStream (IonProcess.java line 65), if I moved this line out of try-with-resource, it works normally. I can also add an extra try..catch for line 70-74 to handle this issues but it may be redundant. Do you have any good idea?

Reply:
I'm confused. If line 71 or 73 throws an IOException, it should definitely be caught by the catch on line 75. This shouldn't have anything to do with line 65. What am I missing here?

Reply:
If I simply change line 70-74 to throw new IOException("test"), which the code will be:
try (
                ......
        ) {
             &nbsp;&nbsp;&nbsp;throw new IOException("test")
        } catch (IOException e) {
            &nbsp;&nbsp;&nbsp;System.err.println("Failed to close OutputStream: " + e.getMessage());
        } catch (IllegalStateException e) {
            &nbsp;&nbsp;&nbsp;System.err.println(e.getMessage());
        }

this part doesn't catch the exception:
**the catch (IOExeption e) {
     &nbsp;&nbsp;&nbsp;System.err.println("Failed to close OutputStream: " + e.getMessage());
}**

Reply:
I just ran this snippet and `"Failed to close OutputStream: test"` does indeed get printed. I think there may be something else going on here.

Reply:
Thanks for testing, I guess I found the reason: can't build an ionWriter with a null outputstream which will throw a NullPointerException.